.travis.yml

```diff
@@ -16,51 +16,65 @@ os:
 go:
 - "1.10"
 - "1.11"
-# - tip
+#- "1.12"
+
 
 env:
-- GORACE="halt_on_error=1"
-  BUILD_VERSION=$(echo ${TRAVIS_COMMIT} | cut -c 1-10)
-  TM_VERSION=v0.27.4
-  FORCE_TM_TEST=1
+  global:
+    - TM_VERSION=v0.27.4
+    - BUILD_VERSION=$(echo ${TRAVIS_COMMIT} | cut -c 1-10)
+    - MAIN_GO_VERSION=1.11
+    - GORACE="halt_on_error=1"
+    - FORCE_TM_TEST=1
 
 install:
 - make deps
 - go get -d github.com/tendermint/tendermint/... ;
   cd ${GOPATH}/src/github.com/tendermint/tendermint ;
   git checkout ${TM_VERSION} ;
   make get_tools && make check && make install ;
-  cd - ;
-  rm -rf ${GOPATH}/src/github.com/tendermint/tendermint
+  cd -;
+
+cache:
+  directories:
+    - $GOPATH
+    - $GOCACHE
 
 script:
-- make lint
-- make build
-- make test
-- make dist
-
-after_script:
-- make cover
-- ./coverage/upload.sh
-
-after_success:
-- if [[ "$TRAVIS_GO_VERSION" == "1.10" ]] && [[ "$TRAVIS_OS_NAME" == "linux" ]] && [[ "$TRAVIS_BRANCH" == "master" ]] && [[ "$TRAVIS_TAG" == "" ]] && [[ "$TRAVIS_PULL_REQUEST_BRANCH" == "" ]]; then
-  docker login -u "$DOCKER_USERNAME" -p "$DOCKER_PASSWORD";
-  docker tag  "iov1/bcpd:${BUILD_VERSION}" "iov1/bcpd:latest" ;
-  docker push "iov1/bcpd:latest";
-  docker tag  "iov1/bnsd:${BUILD_VERSION}" "iov1/bnsd:latest" ;
-  docker push "iov1/bnsd:latest";
-  docker logout;
-  fi
-# build the tagged image
-- if [[ "$TRAVIS_GO_VERSION" == "1.10" ]] && [[ "$TRAVIS_OS_NAME" == "linux" ]] && [[ "$TRAVIS_TAG" != "" ]]; then
-  docker login -u "$DOCKER_USERNAME" -p "$DOCKER_PASSWORD";
-  docker tag  "iov1/bcpd:${BUILD_VERSION}" "iov1/bcpd:$TRAVIS_TAG" ;
-  docker push "iov1/bcpd:$TRAVIS_TAG";
-  docker tag  "iov1/bnsd:${BUILD_VERSION}" "iov1/bnsd:$TRAVIS_TAG" ;
-  docker push "iov1/bnsd:$TRAVIS_TAG";
-  docker logout;
-  fi
+- make lint;
+- make test;
+- if [[ "$TRAVIS_GO_VERSION" == "$MAIN_GO_VERSION" ]]; then
+    make cover;
+    ./coverage/upload.sh;
+  fi;
+- if [[ "$TRAVIS_GO_VERSION" == "$MAIN_GO_VERSION" && "$TRAVIS_OS_NAME" == "linux" ]]; then
```

Comment:
Nice cleanup here, but can you add it back to `after_success`.
Point is not to run it if tests fail

Reply:
If any step fails then further steps are not called.

Here is an example:
- [change that is adding `exit 2`](https://github.com/iov-one/weave/pull/433/commits/e35b43a18121d117d511d57148d3c4f626ad6832) to the build process
- [CI build](https://travis-ci.com/iov-one/weave/jobs/187386755#L550) that does not run anything after `exit 2`